talisker/testing.py

```diff
@@ -0,0 +1,273 @@
+from contextlib import contextmanager
+
+import requests
+
+import talisker.logs
+
+
+class TestHandler(logging.Handler):
+    """Testing handler that records its logs in memory."""
+    def __init__(self, level=logging.NOTSET):
+        super().__init__(level)
+        self.records = []
+        self.lines = []
+
+    def emit(self, record):
+        self.records.append(record)
+        self.lines.extend(self.format(record).split('\n'))
+
+
+class TestLogger(talisker.logs.StructuredLogger):
+    def __init__(self, name='test'):
+        super().__init__(name)
+        handler = TestHandler()
+        handler.setFormatter(talisker.logs.StructuredFormatter())
+        handler.setLevel(logging.NOTSET)
+        self.addHandler(handler)
+        self._test_handler = handler
+
+    @property
+    def records(self):
+        return self._test_handler.records
+
+    @property
+    def lines(self):
+        return self._test_handler.lines
+
+
+def test_logger():
+    return TestLogger()
+
+
+class LogOutput:
+    """A container for log messages output by a Talisker program.
+
+    Parses log lines, and makes them available for inspection in tests.
+    """
+
+    TIMESTAMP = re.compile(r'^\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d.\d\d\dZ')
+
+    def __init__(self, lines):
+        self.logs = list(self.read(lines))
+
+    def read(self, lines):
+        current = []
+        for line in lines:
+            if self.TIMESTAMP.match(line):
+                if current:
+                    yield self.parse(current)
+                current = []
+            current.append(line)
+        if current:
+            yield self.parse(current)
+
+    def parse(self, logs):
+        """Stupid simple logfmt parser"""
+        log = logs[0]
+        trailer = logs[1:]
+        parsed = shlex.split(log)
+        date, time, level, name, msg = parsed[:5]
+        try:
+            extra = dict((v.split('=', 1)) for v in parsed[5:])
+        except ValueError:
+            assert 0, "failed to parse logfmt: " + log
+        return {
+            'ts': date + " " + time,
+            'level': level,
+            'logger': name,
+            'logmsg': msg,
+            'extra': extra,
+            'trailer': trailer,
+        }
+
+    def _compare_strings(self, needle, haystack):
+        return (
+            all(k in haystack for k in needle) and
+            all(needle[k] in haystack[k] for k in needle)
+        )
+
+    def exists(self, **match):
+        extra = match.pop('extra', None)
+        trailer = match.pop('trailer', None)
+        for log in self.logs:
+            strings_match = self._compare_strings(match, log)
+            extra_match = trailer_match = True
+
+            if extra is not None:
+                extra_match = self._compare_strings(extra, log['extra'])
+
+            if trailer is not None:
+                trailer_match = True
+                for line in trailer:
+                    if not any(line in tline for tline in log['trailer']):
```

Comment:
I wonder if the expected trailer lines should be in the correct order? I don't think it matters for now though.